velox/dwio/common/FileSink.h

```diff
@@ -0,0 +1,201 @@
+    return name_;
+  }
+
+  /// Total number of bytes written.
+  virtual uint64_t size() const {
+    return size_;
+  }
+
+  /// Returns true if file sink supports buffering. In the case when buffering
+  /// is not supported, caller need to buffer data to yield optimal write size.
+  virtual bool isBuffered() const {
+    return true;
+  }
+
+  /// Writes a single data buffer.
+  void write(DataBuffer<char> buffer);
+
+  /// Writes data buffers.
+  virtual void write(std::vector<DataBuffer<char>>& buffers) = 0;
+
+  const MetricsLogPtr& metricsLog() const {
+    return metricLogger_;
+  }
+
+  using Factory = std::function<std::unique_ptr<FileSink>(
+      const std::string& name,
+      const FileSink::Options& options)>;
+
+  static bool registerFactory(const Factory& factory);
+
+  static std::unique_ptr<FileSink> create(
+      const std::string& filePath,
+      const Options& options);
+
+ protected:
+  // General write wrapper with logging. All concrete subclasses gets logging
+  // for free if they call a public method that goes through this method.
+  void writeWithLogging(std::vector<DataBuffer<char>>& buffers);
+
+  void writeImpl(
+      std::vector<DataBuffer<char>>& buffers,
+      const std::function<uint64_t(const DataBuffer<char>&)>& callback);
+
+  const std::string name_;
+  memory::MemoryPool* const pool_;
+  const MetricsLogPtr metricLogger_;
+  IoStatistics* const stats_;
+
+  uint64_t size_;
+};
+
+/// Wrapper class that delegates calls to the underlying write file.
+class FileSinkWrapper final : public FileSink {
+ public:
+  FileSinkWrapper(
+      std::unique_ptr<WriteFile> writeFile,
+      std::string name,
+      MetricsLogPtr metricLogger = MetricsLog::voidLog(),
+      IoStatistics* stats = nullptr);
+
+  ~FileSinkWrapper() override {
+    destroy();
+  }
+
+  bool isBuffered() const override {
+    return false;
+  }
+
+  using FileSink::write;
+
+  void write(std::vector<DataBuffer<char>>& buffers) override;
+
+  void doClose() override;
+
+  // TODO: Hack to make Alpha writer work with Velox.  To be removed after Alpha
+  // writer takes DataSink directly.
+  std::unique_ptr<WriteFile> toWriteFile() {
+    markClosed();
+    return std::move(writeFile_);
+  }
+
+ private:
+  std::unique_ptr<WriteFile> writeFile_;
+};
+
+class LocalFileSink : public FileSink {
+ public:
+  LocalFileSink(const std::string& name, const Options& options);
+#ifdef VELOX_ENABLE_BACKWARD_COMPATIBILITY
+  LocalFileSink(const std::string& name, MetricsLogPtr metricLogger)
+      : LocalFileSink(name, {.metricLogger = std::move(metricLogger)}) {}
+#endif
+
+  ~LocalFileSink() override {
+    destroy();
+  }
+
+  using FileSink::write;
+
+  void write(std::vector<DataBuffer<char>>& buffers) override;
+
+  static void registerFactory();
+
+ protected:
+  void doClose() override {
+    ::close(fd_);
+  }
+
+ private:
+  // The local open file handle.
+  int fd_;
+};
+
+class MemorySink : public FileSink {
```

Comment:
@xiaoxmeng I remember we struggled with this naming before due to this class.
Should use `OutputSink` instead of `FileSink`?

Reply:
What about the `MemorySink` class that now extends `FileSink`?
We also seem to be writing to memory?